File: Samples/PurchasePath/Source/Objective-C/PurchasePathExtension/CustomCard/NightOwlCard.m
#import "NightOwlCard.h"
#import "NightOwlCardView.h"
#import "UIColor+HexColors.h"

static NSString * const NightOwlCardRedeemedImageURLString = @"https://developer.usebutton.com/assets/guides/publishers/checkout-extensions/custom-cards/img_owl-redeemed.png";

@interface NightOwlCard ()

@property (nonatomic, assign) BOOL isRedeemed;
@property (nonatomic, strong) UIImage *redeemedImage;

@end

@implementation NightOwlCard

+ (UIView *)createView {
    return [NightOwlCardView nightOwlCardView];
}

- (void)prepareView:(UIView *)cardView {
    NightOwlCardView *view = (NightOwlCardView *)cardView;
    view.titleLabel.text = @"Night Owl Bonus +1%";
    view.bodyLabel.text = @"Whoooo's up late shopping? Here's an extra 1% cash back when you shop now before 6 AM.";
    view.actionButton.layer.cornerRadius = 4.0;
    [view.actionButton setTitle:@"Tap to Hoot" forState:UIControlStateNormal];
    [view.actionButton addTarget:self action:@selector(redeemAction:) forControlEvents:UIControlEventTouchUpInside];
    [self updateView:view];
}

- (void)updateView:(NightOwlCardView *)view {
    view.actionButton.userInteractionEnabled = !_isRedeemed;
    
    
    if (_isRedeemed) {
        view.actionButton.backgroundColor = [UIColor colorWithHex:0x0CAC56];
        [view.actionButton setTitle:@"Offer Redeemed" forState:UIControlStateNormal];
        view.imageView.image = _redeemedImage;
    }
    else {
        view.actionButton.backgroundColor = [UIColor colorWithHex:0x1b84ff];
        [view.actionButton setTitle:@"Tap to Hoot" forState:UIControlStateNormal];
        view.imageView.image = [UIImage imageNamed:@"nightowl"];
    }
}

- (void)redeemAction:(UIButton *)sender {
    NightOwlCardView *view = (NightOwlCardView *)self.view;
    NSError *error = nil;
    NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:NightOwlCardRedeemedImageURLString] options:0 error:&error];
    if (error) {
        NSLog(@"error downloading image: %@", error.localizedDescription);
        return;
    }
    _redeemedImage = [UIImage imageWithData:data scale:[[UIScreen mainScreen] scale]];
    _isRedeemed = YES;
    [self updateView:view];
}

@end

File: Samples/PurchasePath/Source/Swift/PurchasePathExtension/Custom Card/NightOwlCard.swift
import UIKit
import Button

class NightOwlCard: BTNCard {
    
    static let redeemedImageURL = URL(string: "https://developer.usebutton.com/assets/guides/publishers/checkout-extensions/custom-cards/img_owl-redeemed.png")!
    
    var isRedeemed: Bool = false
    var redeemedImage: UIImage?
    
    override static func createView() -> UIView {
        return NightOwlCardView.nightOwlCardView()
    }
    
    override func prepareView(_ cardView: UIView) {
        guard let nightOwlCardView = cardView as? NightOwlCardView else {
            return
        }
        nightOwlCardView.titleLabel.text = "Night Owl Bonus +1%"
        nightOwlCardView.bodyLabel.text = "Whoooo's up late shopping? Here's an extra 1% cash back when you shop now before 6 AM."
        nightOwlCardView.actionButton.layer.cornerRadius = 4
        nightOwlCardView.actionButton.addTarget(self, action: #selector(redeemAction(sender:)), for: .touchUpInside)
        updateView(view: nightOwlCardView)
    }
    
    func updateView(view: NightOwlCardView) {
        view.actionButton.isUserInteractionEnabled = !isRedeemed
        
        if isRedeemed {
            view.actionButton.backgroundColor = UIColor(hex: 0x0CAC56)
            view.actionButton.setTitle("Offer Redeemed", for: .normal)
            view.imageView.image = redeemedImage
        }
        else {
            view.actionButton.backgroundColor = UIColor(hex: 0x1b84ff)
            view.actionButton.setTitle("Tap to Hoot", for: .normal)
            view.imageView.image = UIImage(named: "nightowl")
        }
    }
    
    @objc func redeemAction(sender: Any) {
        guard let cardView = self.view as? NightOwlCardView else {
            return
        }
        
        if let imageData = try? NSData(contentsOf: NightOwlCard.redeemedImageURL, options: NSData.ReadingOptions()) {
            redeemedImage = UIImage(data: imageData as Data, scale: UIScreen.main.scale)?.with(cardView.imageView.bounds.size)
            isRedeemed = true
            updateView(view: cardView)
        }
    }
}
